oqsprov/oqsprov_keys.c

```diff
@@ -33,26 +35,110 @@ void oqsx_freeprovctx(PROV_OQS_CTX *ctx) {
+
+    evp_ctx->kex_info = &nids_ecp[nistlevel - 1];
+
+    evp_ctx->kex = EVP_PKEY_CTX_new_id(evp_ctx->kex_info->nid_kex, NULL);
+    ON_ERR_GOTO(!evp_ctx->kex, err);
+
+    ret = EVP_PKEY_paramgen_init(evp_ctx->kex);
+    ON_ERR_GOTO(ret <= 0, err);
+
+    ret = EVP_PKEY_CTX_set_ec_paramgen_curve_nid(evp_ctx->kex, evp_ctx->kex_info->nid_kex_crv);
+    ON_ERR_GOTO(ret <= 0, err);
+
+    ret = EVP_PKEY_paramgen(evp_ctx->kex, &evp_ctx->kexParam);
+    ON_ERR_GOTO(ret <= 0 || !evp_ctx->kexParam, err);
+
+    err:
+    return ret;
+}
+
+static int oqshybkem_init_ecx(int nistlevel, OQSX_EVP_CTX *evp_ctx)
+{
+    int ret = 1;
+
+    evp_ctx->kex_info = &nids_ecx[nistlevel - 1];
+
+    evp_ctx->kexParam = EVP_PKEY_new();
+    ON_ERR_SET_GOTO(!evp_ctx->kexParam, ret, -1, err);
+
+    ret = EVP_PKEY_set_type(evp_ctx->kexParam, evp_ctx->kex_info->nid_kex);
+    ON_ERR_SET_GOTO(ret <= 0, ret, -1, err);
+
+    evp_ctx->kex = EVP_PKEY_CTX_new(evp_ctx->kexParam, NULL);
+    ON_ERR_SET_GOTO(!evp_ctx->kex, ret, -1, err);
+
+    err:
+    return ret;
+}
+
+static const int (*init_kex_fun[])(int, OQSX_EVP_CTX *) = {
+        oqshybkem_init_ecp,
+        oqshybkem_init_ecx
+};
+
+OQSX_KEY *oqsx_key_new(OSSL_LIB_CTX *libctx, char* oqs_name, char* tls_name, int primitive, const char *propq)
 {
     OQSX_KEY *ret = OPENSSL_zalloc(sizeof(*ret));
+    int ret2 = 0;
 
-    if (ret == NULL)
-        return NULL;
+    if (ret == NULL) goto err;
+
+    if (primitive == KEY_TYPE_SIG) {
+        ret->numkeys = 1;
+        ret->privkey_ptr = OPENSSL_malloc(sizeof(void *));
+        ret->pubkey_ptr = OPENSSL_malloc(sizeof(void *));
+        ret->oqsx_provider_ctx.oqsx_qs_ctx.sig = OQS_SIG_new(oqs_name);
+        ret->privkeylen = ret->oqsx_provider_ctx.oqsx_qs_ctx.sig->length_secret_key;
+        ret->pubkeylen = ret->oqsx_provider_ctx.oqsx_qs_ctx.sig->length_public_key;
+        ret->keytype = KEY_TYPE_SIG;
+    } else if (primitive == KEY_TYPE_KEM) {
+        ret->numkeys = 1;
+        ret->privkey_ptr = OPENSSL_malloc(sizeof(void *));
+        ret->pubkey_ptr = OPENSSL_malloc(sizeof(void *));
+        ret->oqsx_provider_ctx.oqsx_qs_ctx.kem = OQS_KEM_new(oqs_name);
+        ret->privkeylen = ret->oqsx_provider_ctx.oqsx_qs_ctx.kem->length_secret_key;
+        ret->pubkeylen = ret->oqsx_provider_ctx.oqsx_qs_ctx.kem->length_public_key;
+        ret->keytype = KEY_TYPE_KEM;
+    } else if (primitive == KEY_TYPE_ECX_HYB_KEM || primitive == KEY_TYPE_ECP_HYB_KEM) {
+        ret->oqsx_provider_ctx.oqsx_qs_ctx.kem = OQS_KEM_new(oqs_name);
+        ON_ERR_GOTO(!ret->oqsx_provider_ctx.oqsx_qs_ctx.kem, err);
+        OQSX_EVP_CTX *evp_ctx = OPENSSL_zalloc(sizeof(OQSX_EVP_CTX));
+        ON_ERR_GOTO(!evp_ctx, err);
+
+        ret2 = (init_kex_fun[primitive - KEY_TYPE_ECP_HYB_KEM])
+                (ret->oqsx_provider_ctx.oqsx_qs_ctx.kem->claimed_nist_level, evp_ctx);
+        ON_ERR_GOTO(ret2 <= 0 || !evp_ctx->kexParam || !evp_ctx->kex, err);
+
+        ret->numkeys = 2;
```

Comment:
Here'd be another place where iterating the "algorithm array" proposed above would permit creating arbitrary hybrid schemes. Again, not required for this PR.

Reply:
I agree. Similarly as in the comment above, I'd ideally like to combine this with a more uniform API to query key-, secret-, and ciphertext-sizes.